src/tools/isabelle_executor.py

```diff
@@ -466,7 +491,34 @@ def _run_stmt_on_isabelle_server(self, stmt: str) -> None:
                 self.curr_lemma_name, self.curr_lemma = None, ""
                 self.proof_context = None
 
-    def _parse_proof_context(self, proof_context_str: str, local_hypotheses: typing.List[IsabelleLemma]) -> ProofContext:
+    # PISA only supports sledgehammer as an atomic operation. So we must split any tactic which uses it
+    def _handle_sledgehammer(self, start_state: str, step: str, end_state: str) -> str:
```

Comment:
Sledgehammer now works locally, i.e. it can discharge goals from a single line of tactics. Since PISA only supports using sledgehammer atomically, this required some further parsing to break down a line into multiple executable statements when sledgehammer is involved.

To use it, the agent should just issue the sledgehammer command in prove mode. Note that in the Isabelle IDE, this command prints out possible tactics but doesn’t actually apply them. In my implementation, issuing the command will also apply the generated tactics and advance the proof state.